jdaviz/core/template_mixin.py

```diff
@@ -747,6 +760,132 @@ def _selected_changed(self, event):
+
+        def _default_file_parser(path):
+            # by default, just return the file path itself (and allow all files)
+            return '', {path: path}
+
+        self._file_parser = kwargs.pop('file_parser', _default_file_parser)
+        self.add_observe('from_file', self._from_file_changed)
+
+    @property
+    def selected_obj(self):
+        if self.selected == 'From File...':
+            return self._cached_obj.get(self.from_file, self._file_parser(self.from_file)[1])
+        return super().selected_obj
+
+    def _from_file_changed(self, event):
+        if event['new'].startswith('API:'):
+            # object imported from the API: parsing is already handled
+            return
+        if len(event['new']):
+            if event['new'] != self.plugin._file_chooser.file_path:
+                # then need to run the parser or check for valid path
+                if not os.path.exists(event['new']):
+                    if self.selected == 'From File...':
+                        self.select_previous()
+                    raise ValueError(f"{event['new']} is not a valid file path")
+
+                # run through the parsers and check the validity
+                self._on_file_path_changed(event)
+                if self.from_file_message:
+                    if self.selected == 'From File...':
+                        self.select_previous()
+                    raise ValueError(self.from_file_message)
+
+            self.selected = 'From File...'
+
+        elif self.selected == 'From File...':
+            self.select_previous()
+
+    def _on_file_path_changed(self, event):
+        self.from_file_message = 'Checking if file is valid'
+        path = event['new']
+        if (path is not None
+                and not os.path.exists(path)
+                or not os.path.isfile(path)):
+            self.from_file_message = 'File path does not exist'
+            return
+
+        self.from_file_message, self._cached_obj = self._file_parser(path)
+
+    def import_file(self, path):
+        """
+        Select 'From File...' and set the path.
+        """
+        # NOTE: this will trigger self._from_file_changed which in turn will
+        # pass through the parser, raise an error if necessary, and set
+        # self.selected accordingly
+        self.from_file = path
+
+    def import_obj(self, obj):
+        """
+        Import a supported object directly from the API.
+        """
+        msg, self._cached_obj = self._file_parser(obj)
+        if msg:
+            raise ValueError(msg)
+        self.from_file = list(self._cached_obj.keys())[0]
+        self.selected = 'From File...'
+
+
+class HasFileImportSelect(VuetifyTemplate, HubListener):
+    from_file = Unicode().tag(sync=True)
+    from_file_message = Unicode().tag(sync=True)
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        from jdaviz.configs.default.plugins.data_tools.file_chooser import FileChooser
```

Comment:
Why put the import here?

Reply:
I'll double check if this was just lazy testing (in which case I'll move it up) or if it would otherwise cause a circular import (in which case I'll add a comment).

Reply:
it was a circular import - I added a comment. I think "fixing" the problem would require moving data tools and viewers outside of the plugins subdirectory, which is something we might want to consider eventually as tech debt.  This `template_mixin.py` file is also getting large and unorganized and somewhat of a catch-all.... we should probably think about splitting it up at some point as well 😬